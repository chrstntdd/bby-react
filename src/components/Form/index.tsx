import React, { PureComponent } from 'react';

interface IInput {
  id: string;
  value: string;
  isValid?: boolean;
  checked?: boolean;
}

interface PForm {
  handleSubmitCb: () => any;
  refs: any[];
}

interface SForm {
  inputs: [IInput];
}

export class Form extends PureComponent<PForm, SForm> {
  state = {
    fields: {
      ...Object.keys(this.props.fieldDefaults).reduce(
        (acc, curr) => ({
          ...acc,
          [curr]: {
            value: this.props.fieldDefaults[curr],
            isDirty: false
          }
        }),
        {}
      )
    }
  };

  // componentDidUpdate() {
  //   console.log(this.props.refs);
  //   const uh = this.props.refs.map(({ current }) => current.attributes['data-isvalid'].value);
  //   console.log(uh);
  //   // console.dir(this.props.children.state);
  // }

  handleSubmit = e => {
    /* GET THE STATE OF EACH OF THE CHILDREN AND RETURN STATE */
    e.preventDefault();

    this.props.handleSubmitCb();
  };

  handleInputChange = (inputId: any, isValid: boolean, value?: string) => {
    console.log('CALLED');
    this.setState({
      [inputId]: {
        isValid: isValid,
        value: value
      }
    });
  };

  render() {
    return <form onSubmit={this.handleSubmit}>{this.props.children}</form>;
  }
}
