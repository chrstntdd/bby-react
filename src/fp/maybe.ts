export enum MaybeType {
  Nothing,
  Just
}

/* For 'pattern matching' on the caseOf the maybe */
export interface MaybePatterns<T, U> {
  just: (t: T) => U;

  nothing: () => U;
}

/* Public util function to create a Maybe type */
export const maybe = <T>(t: T): Maybe<T> => Maybe.maybe(t);

export type OptionalMaybePatterns<T, U> = Partial<MaybePatterns<T, U>>;

export class Maybe<T> {
  private constructor(private type: MaybeType, private value?: T) {}

  static maybe<T>(t?: T | null): Maybe<T> {
    return t === null || t === undefined
      ? new Maybe<T>(MaybeType.Nothing)
      : new Maybe<T>(MaybeType.Just, t);
  }

  static just<T>(t: T) {
    if (t === null || t === undefined) {
      throw new TypeError('Cannot Maybe.just(null)');
    }
    return new Maybe<T>(MaybeType.Just, t);
  }

  static nothing<T>() {
    return new Maybe<T>(MaybeType.Nothing);
  }

  static isJust<T>(t: Maybe<T>): boolean {
    return t.type === MaybeType.Just;
  }

  static isNothing<T>(t: Maybe<T>): boolean {
    return t.type === MaybeType.Nothing;
  }

  toMaybe<U>(u: U) {
    return Maybe.maybe<U>(u);
  }

  chain = this.bind;
  bind<U>(f: (t: T) => Maybe<U>) {
    return this.type === MaybeType.Just ? f(this.value) : Maybe.nothing<U>();
  }

  caseOf<U>(patterns: MaybePatterns<T, U>) {
    return this.type === MaybeType.Just ? patterns.just(this.value) : patterns.nothing();
  }

  /* Provide a default value, turning an optional value into a normal value. */
  withDefault<U extends T>(defaultValue: U): T | U {
    return this.type === MaybeType.Just ? this.value : defaultValue;
  }

  /* Apply a function to the Maybe */
  lift = this.fMap;
  map = this.fMap;
  fMap<U>(f: (t: T) => U) {
    return this.bind(v => this.toMaybe<U>(f(v)));
  }

  /* Execute a function based on the content of the Maybe. 
     Returns the original value. Meant for side effects!
  */
  do(patterns: Partial<MaybePatterns<T, void>> = {}): Maybe<T> {
    const noop = {
      just: (t: T) => {},
      nothing: () => {}
    };
    const merged = Object.assign(noop, patterns);
    this.caseOf(merged);
    return this;
  }
}
