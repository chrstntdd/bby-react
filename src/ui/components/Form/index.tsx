import React, { Component } from 'react';

import { Maybe, callAll } from '@/fp';
import { debounce } from '@/util';

interface InputField {
  value?: string;
  isValid?: boolean;
  touched?: boolean;
  validationMsg?: string;
  validationFn?: () => (val: string) => Maybe<string>;
}

interface PForm {
  id: string;
  className: string;
  fieldDefaults: (string | InputField)[][];
  onInputChange?: (any) => any;
  onFormSubmit: (any) => any;
  render: (any) => void;
}

interface SForm {
  fields: Map<string, InputField>;
  fieldKeys: string[];
  allFieldsValid: boolean;
}

export class Form extends Component<PForm & React.HTMLProps<HTMLFormElement>, SForm> {
  constructor(props) {
    super(props);

    const state = {
      fields: new Map(this.props.fieldDefaults),
      fieldKeys: this.props.fieldDefaults.map(key => key[0]),
      allFieldsValid: false
    };

    state.fieldKeys.forEach(key => {
      state.fields.set(key, {
        ...state.fields.get(key),
        value: state.fields.get(key).value || '',
        isValid: true,
        touched: false
      });
    });

    this.state = state;
  }

  handleFormSubmit = e => {
    e.preventDefault();

    if (this.state.allFieldsValid) {
      const values = this.state.fieldKeys.map(fieldKey => [
        fieldKey,
        this.state.fields.get(fieldKey).value.trim()
      ]);

      /* RETURNED AS A 2D MATRIX */
      this.props.onFormSubmit(values);
      this.clearFormValues();
    }
  };

  handleInputFocus = e => {
    const fieldKey = e.target.id;

    if (this.state.fields.get(fieldKey).touched === false) {
      this.setState(prevState => {
        prevState.fields.set(fieldKey, {
          ...prevState.fields.get(fieldKey),
          touched: true
        });

        return {
          fields: prevState.fields
        };
      });
    }
  };

  clearFormValues() {
    this.setState(prevState => {
      prevState.fieldKeys.forEach(key => {
        prevState.fields.set(key, {
          ...prevState.fields.get(key),
          value: '',
          isValid: true,
          touched: false
        });
      });

      return {
        fields: prevState.fields
      };
    });
  }

  validateInput = debounce(fieldKey => {
    const field = this.state.fields.get(fieldKey);
    const validationMsg = field.validationFn()(field.value);
    let fieldsToUpdate;

    validationMsg.caseOf({
      /* FIELD IS INVALID */
      just: msg => {
        fieldsToUpdate = {
          isValid: false,
          validationMsg: msg
        };
      },

      /* FIELD IS VALID */
      nothing: () => {
        fieldsToUpdate = {
          isValid: true,
          validationMsg: ''
        };
      }
    });

    this.setState(
      prevState => {
        prevState.fields.set(fieldKey, {
          ...prevState.fields.get(fieldKey),
          ...fieldsToUpdate
        });

        return {
          fields: prevState.fields
        };
      },
      () => this.validateWholeForm()
    );
  }, 400);

  validateWholeForm() {
    this.setState({
      allFieldsValid: this.state.fieldKeys.every(key => {
        const field = this.state.fields.get(key);

        return field.isValid && field.value !== '';
      })
    });
  }

  updateField = event => {
    const target = event.target;
    const fieldKey = target.id;
    const value = target.type === 'checkbox' ? target.checked : target.value;

    this.setState(
      prevState => {
        prevState.fields.set(fieldKey, {
          ...prevState.fields.get(fieldKey),
          value
        });

        return {
          fields: prevState.fields
        };
      },
      () => this.validateInput(fieldKey)
    );
  };

  getInputProps = ({ id, onChange, onFocus, ...rest } = {}) => {
    const field = this.state.fields.get(id);

    return {
      ...rest,
      id,
      value: field.value,
      isValid: field.isValid,
      validationMsg: field.validationMsg,
      className: field.value === '' ? '' : 'has-content',
      onChange: callAll(onChange, this.updateField),
      onFocus: callAll(onFocus, this.handleInputFocus)
    };
  };

  getFormState = () => {
    return {
      allFieldsValid: this.state.allFieldsValid
    };
  };

  getStateAndHelpers() {
    return {
      getInputProps: this.getInputProps,
      getFormState: this.getFormState
    };
  }

  render() {
    return (
      <form className={this.props.className} onSubmit={this.handleFormSubmit}>
        {this.props.render(this.getStateAndHelpers())}
      </form>
    );
  }
}

export default Form;
