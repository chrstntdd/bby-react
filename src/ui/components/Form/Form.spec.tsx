import React from 'react';
import { shallow } from 'enzyme';

jest.useFakeTimers();

import { maybe, Maybe } from '@/fp';

import Form from './';

describe('Form Component', () => {
  const fieldDefaultsMock = [
    [
      'employeeNumberInput',
      {
        validationFn: jest.fn()
      }
    ],
    [
      'passwordInput',
      {
        validationFn: jest.fn()
      }
    ]
  ];

  const requiredProps = {
    id: 'test',
    fieldDefaults: fieldDefaultsMock,
    onFormSubmit: jest.fn(),
    render: jest.fn()
  };

  it('should render without crashing', () => {
    let render;

    beforeAll(() => {
      render = Form.prototype.render;
      Form.prototype.render = jest.fn();
    });

    afterAll(() => {
      Form.prototype.render = render;
    });
    const wrapper = shallow(<Form {...requiredProps} />);

    expect(wrapper).toBeDefined();
  });

  describe('updateField method', () => {
    let render;

    beforeAll(() => {
      render = Form.prototype.render;
      Form.prototype.render = jest.fn();
    });

    afterAll(() => {
      Form.prototype.render = render;
    });
    it('should set the current value of the input to the fieldKey in the Map in state and call validateInput', () => {
      const wrapper = shallow(<Form {...requiredProps} />);
      const validateInputMock = jest.fn();
      const fieldKey = fieldDefaultsMock[0][0];
      const mockEvent = {
        target: {
          id: fieldKey,
          value: 'asdf'
        }
      };

      (wrapper.instance() as Form).validateInput = validateInputMock;

      (wrapper.instance() as Form).updateField(mockEvent);

      expect(wrapper.state().fields.get(fieldKey).value).toBe('asdf');
      expect(validateInputMock).toHaveBeenCalledTimes(1);
      expect(validateInputMock).toHaveBeenCalledWith(fieldKey);
    });

    it('should **NOT** call validateInput method if a validation function does not exist', () => {
      const fieldKey = fieldDefaultsMock[0][0];
      const mockProps = {
        ...requiredProps,
        fieldDefaults: [[fieldKey, {}], requiredProps.fieldDefaults[1]]
      };
      const wrapper = shallow(<Form {...mockProps} />);
      const validateInputMock = jest.fn();
      const mockEvent = {
        target: {
          id: fieldKey,
          value: 'asdf'
        }
      };

      (wrapper.instance() as Form).validateInput = validateInputMock;

      (wrapper.instance() as Form).updateField(mockEvent);

      expect(wrapper.state().fields.get(fieldKey).value).toBe('asdf');
      expect(validateInputMock).toHaveBeenCalledTimes(0);
    });
  });

  describe('handleFormSubmit method', () => {
    let render;

    beforeAll(() => {
      render = Form.prototype.render;
      Form.prototype.render = jest.fn();
    });

    afterAll(() => {
      Form.prototype.render = render;
    });
    describe('when all fields are valid', () => {
      it('should call the onFormSubmit prop callback and clear the form', () => {
        const onFormSubmitMock = jest.fn();
        const clearFormValuesMock = jest.fn();
        const wrapper = shallow(<Form {...requiredProps} onFormSubmit={onFormSubmitMock} />);
        const mockEvent = {
          preventDefault: jest.fn()
        };

        (wrapper.instance() as Form).clearFormValues = clearFormValuesMock;
        wrapper.update();

        wrapper.setState({
          allFieldsValid: true
        });

        (wrapper.instance() as Form).handleFormSubmit(mockEvent);

        expect(onFormSubmitMock).toHaveBeenCalledTimes(1);
        expect(onFormSubmitMock).toHaveBeenCalledWith([
          ['employeeNumberInput', ''],
          ['passwordInput', '']
        ]);
        expect(clearFormValuesMock).toHaveBeenCalledTimes(1);
      });
    });

    describe('when all fields are not valid', () => {
      it('should do nothing', () => {
        const onFormSubmitMock = jest.fn();
        const clearFormValuesMock = jest.fn();
        const wrapper = shallow(<Form {...requiredProps} onFormSubmit={onFormSubmitMock} />);
        const mockEvent = {
          preventDefault: jest.fn()
        };

        (wrapper.instance() as Form).clearFormValues = clearFormValuesMock;
        wrapper.update();

        wrapper.setState({
          allFieldsValid: false
        });

        (wrapper.instance() as Form).handleFormSubmit(mockEvent);

        expect(onFormSubmitMock).toHaveBeenCalledTimes(0);
        expect(clearFormValuesMock).toHaveBeenCalledTimes(0);
      });
    });
  });

  describe('handleInputFocus method', () => {
    let render;

    beforeAll(() => {
      render = Form.prototype.render;
      Form.prototype.render = jest.fn();
    });

    afterAll(() => {
      Form.prototype.render = render;
    });
    describe('when the field has not been touched', () => {
      it('should set touched to true', () => {
        const wrapper = shallow(<Form {...requiredProps} />);
        const fieldKey = fieldDefaultsMock[0][0];
        const mockEvent = {
          target: {
            id: fieldKey
          }
        };

        (wrapper.instance() as Form).handleInputFocus(mockEvent);

        expect(wrapper.state().fields.get(fieldKey).touched).toBeTruthy();
      });
    });

    describe('when the field has already been touched', () => {
      it('should do nothing', () => {
        const wrapper = shallow(<Form {...requiredProps} />);
        const initialState = wrapper.state();
        const fieldKey = fieldDefaultsMock[0][0];
        const mockEvent = {
          target: {
            id: fieldKey
          }
        };

        wrapper.setState(prevState => {
          prevState.fields.set(fieldKey, {
            ...prevState.fields.get(fieldKey),
            touched: true
          });

          return {
            fields: prevState.fields
          };
        });

        (wrapper.instance() as Form).handleInputFocus(mockEvent);

        expect(wrapper.state()).toEqual(initialState);
      });
    });
  });

  describe('clearFormValues method', () => {
    let render;

    beforeAll(() => {
      render = Form.prototype.render;
      Form.prototype.render = jest.fn();
    });

    afterAll(() => {
      Form.prototype.render = render;
    });
    it('should do its job', () => {
      const wrapper = shallow(<Form {...requiredProps} />);
      const initialState = wrapper.state();

      wrapper.setState(prevState => {
        prevState.fieldKeys.forEach(key => {
          prevState.fields.set(key, {
            ...prevState.fields.get(key),
            value: 'duplicate',
            isValid: false,
            touched: true
          });
        });

        return {
          fields: prevState.fields
        };
      });

      (wrapper.instance() as Form).clearFormValues();

      expect(wrapper.state()).toEqual(initialState);
    });
  });

  describe('validateInput method', () => {
    let render;
    const fieldKey = fieldDefaultsMock[0][0];
    const validateWholeFormMock = jest.fn();

    beforeAll(() => {
      render = Form.prototype.render;
      Form.prototype.render = jest.fn();
    });

    beforeEach(() => {
      validateWholeFormMock.mockClear();
    });

    afterAll(() => {
      Form.prototype.render = render;
    });
    describe('when validationMsg is Just a validationMsg', () => {
      it('should set isValid to false and set the message into the state', () => {
        const mockProps = {
          ...requiredProps,
          fieldDefaults: [
            [
              requiredProps.fieldDefaults[0][0],
              { validationFn: () => jest.fn(() => maybe('ERROR')) }
            ],
            requiredProps.fieldDefaults[1]
          ]
        };

        const wrapper = shallow(<Form {...mockProps} />);

        (wrapper.instance() as Form).validateWholeForm = validateWholeFormMock;
        wrapper.update();

        (wrapper.instance() as Form).validateInput(fieldKey);

        jest.runOnlyPendingTimers();

        expect(wrapper.state().fields.get('employeeNumberInput').validationMsg).toEqual('ERROR');
        expect(wrapper.state().fields.get('employeeNumberInput').isValid).toBe(false);
        expect(validateWholeFormMock).toHaveBeenCalledTimes(1);
      });
    });
    describe('when the field is valid', () => {
      it('should set isValid to true and clear the validationMsg', () => {
        const mockProps = {
          ...requiredProps,
          fieldDefaults: [
            [
              requiredProps.fieldDefaults[0][0],
              { validationFn: () => jest.fn(() => Maybe.nothing()) }
            ],
            requiredProps.fieldDefaults[1]
          ]
        };

        const wrapper = shallow(<Form {...mockProps} />);

        (wrapper.instance() as Form).validateWholeForm = validateWholeFormMock;
        wrapper.update();

        (wrapper.instance() as Form).validateInput(fieldKey);

        jest.runOnlyPendingTimers();

        expect(wrapper.state().fields.get('employeeNumberInput').validationMsg).toEqual('');
        expect(wrapper.state().fields.get('employeeNumberInput').isValid).toBe(true);
        expect(validateWholeFormMock).toHaveBeenCalledTimes(1);
      });
    });
  });

  describe('validateWholeForm method', () => {
    let render;

    beforeAll(() => {
      render = Form.prototype.render;
      Form.prototype.render = jest.fn();
    });

    afterAll(() => {
      Form.prototype.render = render;
    });
    it('should do its job', () => {
      const wrapper = shallow(<Form {...requiredProps} />);

      wrapper.setState(prevState => {
        prevState.fieldKeys.forEach(key => {
          prevState.fields.set(key, {
            ...prevState.fields.get(key),
            value: 'validtext',
            isValid: true,
            touched: true
          });
        });

        return {
          fields: prevState.fields
        };
      });

      (wrapper.instance() as Form).validateWholeForm();

      expect(wrapper.state().allFieldsValid).toBe(true);
    });
  });

  describe('getInputProps getter', () => {
    const fieldKey = fieldDefaultsMock[0][0];
    it('should return props to be spread on a form input', () => {
      const form = new Form({ ...requiredProps });

      expect(form.getInputProps({ id: fieldKey.toString() })).toMatchSnapshot();
    });
  });

  describe('getFormState getter', () => {
    it('should return form state', () => {
      const form = new Form({ ...requiredProps });

      expect(form.getFormState()).toMatchSnapshot();
    });
  });

  describe('getStateAndHelpers getter', () => {
    it('should return an object for use in the render function', () => {
      const form = new Form({ ...requiredProps });

      expect(form.getStateAndHelpers()).toMatchSnapshot();
    });
  });
});
