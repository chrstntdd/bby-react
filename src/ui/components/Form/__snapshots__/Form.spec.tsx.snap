// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Component getFormState getter should return form state 1`] = `
Object {
  "allFieldsValid": false,
}
`;

exports[`Form Component getInputProps getter should return props to be spread on a form input 1`] = `
Object {
  "className": "",
  "id": "employeeIdInput",
  "isValid": true,
  "onBlur": [Function],
  "onChange": [Function],
  "onFocus": [Function],
  "validationMsg": undefined,
  "value": "",
}
`;

exports[`Form Component getStateAndHelpers getter should return an object for use in the render function 1`] = `
Object {
  "getFormState": [Function],
  "getInputProps": [Function],
}
`;
